Learner 842 Add ability to show language selector in LMS footer

amangano-edx opened the pull request:
Adds the ability to display a language selector in the footer on the LMS.

Depends on https://github.com/edx/edx-platform/pull/15097

@edx/devops please review the config changes proposed in this PR when you have a chance.

andy-armstrong commented on common/djangoapps/pipeline_mako/templates/static_content.html:
```diff
@@ -199,6 +199,13 @@
     return get_value('SHOW_LANGUAGE_SELECTOR', settings.FEATURES.get('SHOW_LANGUAGE_SELECTOR', False))
 %></%def>
 
+<%def name="show_footer_language_selector(api_setting)"><%
```
I'd rather not put an edX-specific piece of logic into this very general class. Is there a footer.py that it could live in?

andy-armstrong replied:
What is the use case for the ``api_setting`` parameter? My guess is that you want to provide this value when rendering the footer via the API (i.e. the marketing site) and that you want to use the setting when rendering the footer inline. My recommendation would be to have that logic in the clients that are rendering the template so then you wouldn't need this method at all. You would then leave the template rendering based on the context parameter.

amangano-edx replied:
@andy-armstrong I'm not sure that this is any more edx-specific than the `show_language_selector` def. The footer language selector can be used by any openedx instance. All footer templates support it.

The `api_setting` parameter allows us to override what's in settings when requesting the footer via the API, so that we may explicitly decide to include/exclude it on certain pages regardless of what we're doing for the broader LMS.

andy-armstrong replied:
I didn't realize that ``show_language_selector`` was already in this file. This is an inappropriate place for that method too, IMO! I also don't like that the two methods have different signatures, but I suppose that's because we don't render the header via an API.

I feel that this method is too confusing because it is serving different purposes, and the contract has to understand them. How would you write a unit test for this method? Why should the template have to know whether it is being rendered by an API or not? The API only renders in one method, so it seems cleaner to have that pass an extra parameter, rather than having this method need to understand about the API's requirements.

This is one of the unfortunate things about Mako, which is that it lets you evaluate arbitrary Python, so the code gets much more confusing. If this was a Django Template, it would receive the context parameter and just render based upon it. It wouldn't have to know about Django settings or API values.

I wonder if it would be cleaner to have this defaulting logic applied here:

https://github.com/edx/edx-platform/blob/master/common/djangoapps/edxmako/shortcuts.py#L138

Then the templates could rely upon this parameter always being passed through.

Sorry for the philosophical discussion. Let me know if any of this makes sense. Maybe I should just let this go since this code is already confusing, but I'd rather not add to the complexity if we can avoid it.

andy-armstrong commented on lms/envs/common.py:
```diff
@@ -345,9 +345,12 @@
     'SHOW_LANGUAGE_SELECTOR': False,
 
+    # Show language selector in footer.
+    'SHOW_FOOTER_LANGUAGE_SELECTOR': False,
```
I'd rather not have two settings like this that are so closely related but that aren't named consistently. I'd rather change the existing setting to support enum values of ``True``, ``FOOTER_ONLY``, ``HEADER_ONLY`` and ``False``.

amangano-edx replied:
@andy-armstrong I'm a little hesitant to change the existing setting as I fear doing so would not be backwards compatible and may unintentionally break things for upgrading openedx users. I'm happy to change this, but I want to make sure that we do it in a way that will not cause problems for people. Do you have any suggestions for how this could be done?

andy-armstrong replied:
My proposal is that the setting continues to take the two values it does today (i.e. False and True) and that you just support two additional values. Nobody will be using those new values yet, so they shouldn't cause any trouble. I doubt that anyone is coding against these settings so they shouldn't be surprised by the additional values, and if they are then they will be truthy so they will behave as if the value were True.

amangano-edx replied:
I think it would be unexpected if the meaning of the `SHOW_LANGUAGE_SELECTOR: True` changed from 'show the language selector in the header' to 'show the language selector in both the header and footer'. Would it be acceptable instead to create a new setting called `LANGUAGE_SELECTOR`, which could take an enum like you have described, and deprecate the old `SHOW_LANGUAGE_SELECTOR` setting? Do we have a deprecation policy for changes like these?

amangano-edx replied:
@andy-armstrong does this actually make sense as a setting? Would it make more sense to toggle it via waffle switch instead?

amangano-edx commented:
@andy-armstrong I just pushed a WIP commit that I think should address some of the issues that you brought up earlier. In it, I've deprecated the `SHOW_LANGUAGE_SELECTOR` setting and introduced a more generic setting, `LANGUAGE_SELECTOR`, that can be used to configure any language selector in the application. I also refactored the `static_content.html` namespace, splitting out the i18n-specific stuff into it's own namespace that can be included where needed.

Please let me know if this approach is acceptable to you when you have a chance.

amangano-edx commented:
@bderusha does the `LANGUAGE_SELECTOR` setting need to be in edx/configuration? The `SHOW_LANGUAGE_SELECTOR` is not included there? If it is missing, would we just use the default in the `settings/common` file?

andy-armstrong commented on common/djangoapps/pipeline_mako/templates/i18n_helpers.html:
```diff
@@ -0,0 +1,31 @@
+<%page expression_filter="h"/>
```
Thanks, @amangano-edx. It is much better to keep the i18n stuff together. 

Nit: I wonder why this needs to be in ``pipeline_mako``. This doesn't seem like pipeline-specific logic, so I wouldn't expect to see it live here. It seems that it would be simpler if this lived in i18n_helpers.py that had unit tests, rather than having the code be Mako-specific. It would also be great to have all the i18n logic live in one Django app, rather than be spread around the platform.  This is just a nit so you can choose to address it or not.

amangano-edx replied:
@andy-armstrong I actually wasn't sure where this should live, so I'd be happy to move it somewhere else. Do you have any suggestions that would be appropriate? Maybe the lang_pref app?

amangano-edx replied:
Also, the `language_selector_config` def is mako-specific, as it reads the mako context and overrides the `LANGUAGE_SELECTOR` setting if it is present.

andy-armstrong replied:
@amangano-edx that's a good reason to just go with what you have. Thanks!

andy-armstrong commented on cms/envs/bok_choy.env.json:
```diff
@@ -75,7 +75,9 @@
         "ENABLE_CONTENT_LIBRARIES": true,
         "ENABLE_SPECIAL_EXAMS": true,
-        "SHOW_LANGUAGE_SELECTOR": true,
+        "LANGUAGE_SELECTOR": {
```
Thank you for this! This will be much more flexible, and easier to understand.

jibsheet replied:
Please do not add dictionaries to the feature flag list, feature flags should be flags (boolean settings).

bderusha commented:
@amangano-edx I don't think the config is required in the configuration repo given that it isn't there right now and the complexity around getting it working there.  Just something to think about as a future enhancement.  Shouldn't block this PR

jibsheet commented on lms/envs/devstack_docker.py:
```diff
@@ -30,8 +30,10 @@
     'ENABLE_COURSEWARE_SEARCH': False,
     'ENABLE_COURSE_DISCOVERY': False,
     'ENABLE_DASHBOARD_SEARCH': False,
-    'SHOW_LANGUAGE_SELECTOR': True,
```
Why are we deleting this in one place but not others?
Are you ready to remove/deprecate it?

amangano-edx replied:
Yes, the idea is that this setting should be deprecated in favor of the new LANGUAGE_SELECTOR setting. Thinking about it, I can probably just remove it from the other config files as well, so long as the code continues to honor the original setting if it is set on an older installation.

amangano-edx commented:
@jibsheet should I wire this into the aws.py and add a PR to the configuration repo?

jibsheet commented:
@amangano-edx The basic question is: Does the value in common.py work, or do you need to change it for production?  If the latter, you need to wire it into aws.py and configuration.git and override it in edx-internal.

amangano-edx commented on openedx/core/djangoapps/lang_pref/api.py:
```diff
@@ -14,6 +16,20 @@
+
+    # The SHOW_LANGUAGE_SELECTOR setting is deprecated, but might still be in use on some installations.
+    deprecated_setting = get_value('SHOW_LANGUAGE_SELECTOR', settings.FEATURES.get('SHOW_LANGUAGE_SELECTOR', False))
+    return setting or deprecated_setting
```
Here, I believe it makes sense to return the new setting if it has been set to True, and otherwise fall back to the deprecated setting. That way, if the deprecated setting is explicitly set to True somewhere we will honor it. If anyone explicitly uses the new setting to configure the language selector to something other than the default (false), they should take care to remove all references to the old setting. For example, if they update lms.env.py and replace  `'SHOW_LANGUAGE_SELECTOR':True` with `'SHOW_HEADER_LANGUAGE_SELECTOR':True`, they should also take care to update any site-specific configuration referencing `'SHOW_LANGUAGE_SELECTOR'` @andy-armstrong , @jibsheet does this seem reasonable to you?

andy-armstrong replied:
👍  from me.

andy-armstrong replied:
I also like where this helper method landed in the end.

jibsheet replied:
The only other thing is to maybe log a warning "honoring deprecated setting SHOW_LANGUAGE_SELECTOR see X for new settings" but I otherwise like the backcompat work.

amangano-edx commented:
@andy-armstrong @jibsheet I have refactored to use two boolean settings instead of a dictionary. When you have a minute could you please take one final look?

amangano-edx commented:
@jibsheet I just added a log message, per your suggestion. Would you mind taking one more look?

amangano-edx commented on openedx/core/djangoapps/lang_pref/api.py:
```diff
@@ -2,17 +2,41 @@
+    if deprecated_setting is None:
+        deprecated_setting = False
+    else:
+        log.warn("The SHOW_LANGUAGE_SELECTOR setting is deprecated. Use SHOW_HEADER_LANGUAGE_SELECTOR instead.")
```
@jibsheet Should we be concerned about this spamming the log for users that haven't updated their settings? I believe this line would execute on every page since it is part of the header.

jibsheet replied:
Fair point.  I think the fix is an app config check which may not be worth it

amangano-edx replied:
I will remove this to be on the safe side.
